torch/lib/THCUNN/generic/SpatialDepthwiseConvolution.cu

```diff
@@ -0,0 +1,201 @@
+                  int dW, int dH,
+                  int padW, int padH,
+                  int dilationW, int dilationH)
+{
+  THCUNN_assertSameGPU(state, 3, input, output, weight);
+
+  // Only handle 4D Input Tensors for now
+  assert(THCTensor_(nDimension)(state, input) == 4);
+  assert(THCTensor_(nDimension)(state, weight) == 4);
+
+  // We assume that the input and weight Tensors are shaped properly by
+  // the caller, so we verify that here to some extent
+
+  // Weight Tensor is shape (output_channels, 1, kH, kW)
+  assert(weight->size[1] == 1);
+
+  // Input Tensor is shape (N, input_channels, H, W)
+  // We verify that the # of output_channels is a multiple of input_channels
+  assert(weight->size[0] % input->size[1] == 0);
+
+  // Bias has same # of channels as output
+  if (bias) {
+    assert(bias->size[0] == weight->size[0]);
+  }
+
+  // Following the behvaior of other THCUNN functions, we shape the output
+  // Tensor ourselves
+
+  int batchSize = input->size[0];
+  int height = input->size[2];
+  int width = input->size[3];
+  int outputHeight = (height + 2 * padH - (dilationH * (kH - 1) + 1)) / dH + 1;
+  int outputWidth = (width + 2 * padW - (dilationW * (kW - 1) + 1)) / dW + 1;
+  int outputChannels = weight->size[0];
+
+  THCTensor_(resize4d)(state, output, batchSize, outputChannels, outputHeight, outputWidth);
+
+  THCDeviceTensor<real, 4> dInput = toDeviceTensor<real, 4>(state, input);
+  THCDeviceTensor<real, 4> dWeight = toDeviceTensor<real, 4>(state, weight);
+  THCDeviceTensor<real, 4> dOutput = toDeviceTensor<real, 4>(state, output);
+  THCDeviceTensor<real, 1> dBias;
+  if (bias) {
+    dBias = toDeviceTensor<real, 1>(state, bias);
+  }
+
+  // Kernel currently relies upon all the Tensors to be contiguous
+  assert(dInput.isContiguous());
+  assert(dWeight.isContiguous());
+  assert(dOutput.isContiguous());
+
+  int inputChannels = input->size[1];
+  int depthwiseMultiplier = outputChannels / inputChannels;
+
+  // One thread per output value
+  int n = THCTensor_(nElement)(state, output);
+  int blocks = GET_BLOCKS(n);
+  dim3 grid(blocks);
+  dim3 block(CUDA_NUM_THREADS);
+
+  spatialDepthwiseConvolutionUpdateOutput<<<grid, block, 0, THCState_getCurrentStream(state)>>>(
+    dInput, dOutput, dWeight, dBias, bias != NULL, n, outputChannels, depthwiseMultiplier,
+    width, height, outputWidth, outputHeight,
+    kW, kH, dW, dH, padW, padH, dilationW, dilationH);
+
+  THCudaCheck(cudaGetLastError());
+}
+
+void THNN_(SpatialDepthwiseConvolution_updateGradInput)(
+                  THCState *state,
+                  THCTensor *input,
+                  THCTensor *gradOutput,
+                  THCTensor *gradInput,
+                  THCTensor *weight,
+                  int kW, int kH,
+                  int dW, int dH,
+                  int padW, int padH,
+                  int dilationW, int dilationH)
+{
+  THCUNN_assertSameGPU(state, 4, input, gradOutput, gradInput, weight);
```

Comment:
input is not used in kernel, thus does not have to be on the same GPU. Also, same GPU asserts and shape checkes are the same in all 3 functions, save for argument names (updateOutput, updateGradInput, accGradParameters), so can be put into a separate helper function.

Reply:
Good catch. I was playing around with the input during development. However, I'm not a fan of mimic-ing the existing THNN functions which have these shared shape-checks, where you have to scroll back and forth to see what parameter maps to what name in the shape check. My hope is that shape-checking is largely starting to occur at the Python (or ATen) level with more sane error messages. When you get a `THNN` assertion failure, its essentially useless to the end user unless they look into the code themselves.